Update the changelog and documentation (#1033)
PennyLaneAI/catalyst: doc/changelog.md

  This new functionality fits the data from noise-scaled circuits with an exponential function,
  and returns the zero-noise value:

  ```py
  from pennylane.transforms import exponential_extrapolate
  from catalyst import mitigate_with_zne

  dev = qml.device("lightning.qubit", wires=2, shots=100000)

  @qml.qnode(dev)
  def circuit(weights):
      qml.StronglyEntanglingLayers(weights, wires=[0, 1])
      return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))

  @qjit
  def workflow(weights, s):
      zne_circuit = mitigate_with_zne(circuit, scale_factors=s, extrapolate=exponential_extrapolate)
      return zne_circuit(weights)
  ```

  ```pycon
  >>> weights = jnp.ones([3, 2, 3])
  >>> scale_factors = jnp.array([1, 2, 3])
  >>> workflow(weights, scale_factors)
  Array(-0.19946598, dtype=float64)
  ```

* A new module is available, `catalyst.passes`, which provides Python decorators
  for enabling and configuring Catalyst MLIR compiler passes.
  [(#911)](https://github.com/PennyLaneAI/catalyst/pull/911)

  The first pass available is `catalyst.passes.cancel_inverses`,
  which enables the `-removed-chained-self-inverse` MLIR pass that
  cancels two neighbouring Hadamard gates.

>   ```python

Author: Have simplified the example for the changelog

Reviewer: Thanks, this is indeed much cleaner!

  from catalyst.debug import get_compilation_stage
  from catalyst.passes import cancel_inverses

  dev = qml.device("lightning.qubit", wires=1)

  @qml.qnode(dev)
  def circuit(x: float):
      qml.RX(x, wires=0)
      qml.Hadamard(wires=0)
      qml.Hadamard(wires=0)
      return qml.expval(qml.PauliZ(0))

  @qjit(keep_intermediate=True)
  def workflow(x):
      optimized_circuit = cancel_inverses(circuit)

>       return circuit(x), optimized_circuit(x)

Author: Note: this needs #1032 to be solved for this to work correctly

Reviewer: #1032 is done now

Reviewer: In fact I think this is the exact test case I added in #1037

Author: thanks!

  Array(48, dtype=int64)
  ```

* Mid-circuit measurement results when using `lightning.qubit` and `lightning.kokkos`
  can now be seeded via the new `seed` argument of the `qjit` decorator.
  [(#936)](https://github.com/PennyLaneAI/catalyst/pull/936)

>   The seed argument accepts an unsigned 32-bit integer.

Reviewer 2: ```suggestion
  The seed argument accepts an unsigned 32-bit integer, which is used to initialize the pseudo-random
  state at the beginning of each execution of the compiled function.
```

>   Different `qjit` objects with the same seed (including repeated calls to the same `qjit`)

Reviewer 2: ```suggestion
  Therefor, different `qjit` objects with the same seed (including repeated calls to the same `qjit`)
```

      m = measure(0)

      if m:
          qml.Hadamard(0)

      return qml.probs()

  @qjit(seed=37, autograph=True)
  def workflow(x):
      return jnp.stack([circuit(x) for i in range(4)])
  ```

  Repeatedly calling the `workflow` function above will always
  result in the same values:

  ```pycon
  >>> workflow(1.8)
  Array([[1. , 0. ],
       [1. , 0. ],
       [1. , 0. ],
       [0.5, 0.5]], dtype=float64)
  >>> workflow(1.8)
  Array([[1. , 0. ],
       [1. , 0. ],
       [1. , 0. ],
       [0.5, 0.5]], dtype=float64)
  ```

  Note that setting the seed will *not* avoid shot-noise stochasticity in terminal measurement
  statistics such as `sample` or `expval`:

  ```python
  dev = qml.device("lightning.qubit", wires=1, shots=10)

  @qml.qnode(dev)
  def circuit(x):
      qml.RX(x, wires=0)
      m = measure(0)

      if m:
          qml.Hadamard(0)

      return qml.expval(qml.PauliZ(0))

  @qjit(seed=37, autograph=True)
  def workflow(x):
      return jnp.stack([circuit(x) for i in range(4)])
  ```
  ```pycon
  >>> workflow(1.8)
  Array([1. , 1. , 1. , 0.4], dtype=float64)
  >>> workflow(1.8)
  Array([ 1. ,  1. ,  1. , -0.2], dtype=float64)
  ```

> * Exponential extrapolation is now a supported method of zero-noise extrapolation when performing

Author: good call, I also struggled with the repeated extrapolation!

  4.0
  ```

  Here we use `get_compilation_stage` to acquire the IR, and then modify
  `%2 = arith.mulf %in, %in_0 : f64` to turn the square function into a cubic one
  via `replace_ir`:

  ```python
  from catalyst.debug import get_compilation_stage, replace_ir

  old_ir = get_compilation_stage(f, "HLOLoweringPass")
  new_ir = old_ir.replace(
      "%2 = arith.mulf %in, %in_0 : f64\n",
      "%t = arith.mulf %in, %in_0 : f64\n    %2 = arith.mulf %t, %in_0 : f64\n"
  )
  replace_ir(f, "HLOLoweringPass", new_ir)
  ```

  The recompilation starts after the given checkpoint stage:

  ```pycon
  >>> f(2.0)
  8.0
  ```

<h3>Improvements</h3>

* Catalyst has been updated to support JAX v0.4.28.
  [(#931)](https://github.com/PennyLaneAI/catalyst/pull/931)
  [(#995)](https://github.com/PennyLaneAI/catalyst/pull/995)

* Catalyst now supports keyword arguments for qjit-compiled functions.
  [(#1004)](https://github.com/PennyLaneAI/catalyst/pull/1004)

  ```pycon
  >>> @qjit
  ... @grad
  ... def f(x, y):
  ...     return x * y
  >>> f(3., y=2.)
  Array(2., dtype=float64)
  ```

  Note that the `static_argnums` argument to the `qjit` decorator

>   is not supported when passing argument values as keyword arguments.

Author: Something we should add to the backlog!

Reviewer 2: Is this supported now that we fixed keyword argument support in #1004 ?

Author: yep!

  4.0
  ```

  Here we use `get_compilation_stage` to acquire the IR, and then modify
  `%2 = arith.mulf %in, %in_0 : f64` to turn the square function into a cubic one
  via `replace_ir`:

  ```python
  from catalyst.debug import get_compilation_stage, replace_ir

  old_ir = get_compilation_stage(f, "HLOLoweringPass")
  new_ir = old_ir.replace(
      "%2 = arith.mulf %in, %in_0 : f64\n",
      "%t = arith.mulf %in, %in_0 : f64\n    %2 = arith.mulf %t, %in_0 : f64\n"
  )
  replace_ir(f, "HLOLoweringPass", new_ir)
  ```

  The recompilation starts after the given checkpoint stage:

  ```pycon
  >>> f(2.0)

>   8.0

Author: https://github.com/PennyLaneAI/catalyst/pull/1033/files#r1729132298

  ```

> * Catalyst now supports c executable generation with `catalyst.debug.compile_executable`.

Author: Will do!

> * Catalyst now supports c executable generation with `catalyst.debug.compile_executable`.

Author: Done in fab34925

  4.0
  ```

  Here we use `get_compilation_stage` to acquire the IR, and then modify
  `%2 = arith.mulf %in, %in_0 : f64` to turn the square function into a cubic one
  via `replace_ir`:

  ```python
  from catalyst.debug import get_compilation_stage, replace_ir

  old_ir = get_compilation_stage(f, "HLOLoweringPass")
  new_ir = old_ir.replace(
      "%2 = arith.mulf %in, %in_0 : f64\n",
      "%t = arith.mulf %in, %in_0 : f64\n    %2 = arith.mulf %t, %in_0 : f64\n"
  )
  replace_ir(f, "HLOLoweringPass", new_ir)
  ```

  The recompilation starts after the given checkpoint stage:

  ```pycon
  >>> f(2.0)

>   8.0
>   ```

Author: ````suggestion
  8.0
  ```

  Either function can also be used independently of each other.
````

  8.0
  ```

  Either function can also be used independently of each other. Note that
  `get_compilation_stage` replaces the `print_compilation_stage` function;
  please see the Breaking Changes section for more details.

> * Catalyst now supports direct C executable generation with `catalyst.debug.compile_executable`.

Reviewer 2: This is very minor, but I don't know that `C executable` is too relevant because an executable is "written" in machine code, the fact that we generated it using the C language might not be too relevant? 😅

Author: Thank you, I wanted to rewrite this but wasn't sure what the best wording was! anything in mind?

Reviewer 2: Something like this? :) 
```suggestion
* Catalyst now supports generating executables from compiled functions for the native host architecture using
  `catalyst.debug.compile_executable`.
```

Author: Perfect

<h3>Improvements</h3>

> * Catalyst has been updated to support JAX v0.4.28.

Author: Good point

  Array([0.99500417, 0.98006658, 0.95533649], dtype=float64)
  ```

* Verification is now performed before compilation to confirm that the measurements included in QNodes
  are compatible with the specified device and settings.
  [(#945)](https://github.com/PennyLaneAI/catalyst/pull/945)
  [(#962)](https://github.com/PennyLaneAI/catalyst/pull/962)

  ```pycon
  >>> dev = qml.device("lightning.qubit", wires=2, shots=None)
  >>> @qjit
  ... @qml.qnode(dev)
  ... def circuit(params):
  ...     qml.RX(params[0], wires=0)
  ...     qml.RX(params[1], wires=1)
  ...     return {
  ...         "sample": qml.sample(wires=[0, 1]),
  ...         "expval": qml.expval(qml.PauliZ(0))
  ...     }
  >>> circuit([0.1, 0.2])
  CompileError: Sample-based measurements like sample(wires=[0, 1])
  cannot work with shots=None. Please specify a finite number of shots.

* On devices that support it, initial state preparation routines `qml.StatePrep` and `qml.BasisState`
  are no longer decomposed when using Catalyst, improving compilation and runtime performance.
  [(#955)](https://github.com/PennyLaneAI/catalyst/pull/955)

* Catalyst is now compatible with Enzyme `v0.0.130`

>   [(#898)](https://github.com/PennyLaneAI/catalyst/pull/898)

Author: @rmoyard are there any user benefits I can mention here? Or would this best fit under 'internal changes'?

Reviewer 3: You can move it in internal changes :+1:

* Improve error messaging for `catalyst.jvp` when the callee input type and the tangent
  type are not compatible by performing type-checking at the MLIR level. Note that the
  equivalent type checking is already performed in `catalyst.vjp`.
  [(#1020)](https://github.com/PennyLaneAI/catalyst/pull/1020)

>   [(#1030)](https://github.com/PennyLaneAI/catalyst/pull/1030)

Author: @joeycarter is there a user-facing code example I could show here?

Reviewer 4: Not really, no. We're putting together another PR (#1031) with similar type checking in the frontend `catalyst.jvp`. There, a user will see a new error message if incorrect types are given; I'll add an example to the changelog in that PR. However, a user would only see these MLIR error message if they're processing the raw IR or using a different frontend. The reasoning for putting the type checking in both places was for better future-proofing in case we (or someone else) develop a new frontend, where we would then have the MLIR-level type checking as the fail-safe.

* Eliminate (some) scalar tensors from the IR by adding a `linalg-detensorize` pass at the end of the HLO lowering passes.

>   [(#1010)](https://github.com/PennyLaneAI/catalyst/pull/1010)

Author: @dime10 @erick-xanadu would this best fit under 'internal changes'?

Reviewer 2: It could, but it should have performance implications for the user. A reword should point this out.
Too bad we haven't measured the difference.

Reviewer 2: Wording is now updated, but we may still want to move this from the international changes to improvements?

Author: Yep, let's move this to improvements

Author: done in cabdf468

* Adds `catalyst.from_plxpr.from_plxpr` for converting a PennyLane variant jaxpr into a
  Catalyst variant jaxpr.

>   [(#837)](https://github.com/PennyLaneAI/catalyst/pull/837)

Author: I'm not sure what to do here, since this function does not appear in the documentation 🤔 I think it best goes into internal changes.

Reviewer 2: Agreed.

  qjit-compiled function being NumPy arrays will need to be updated.
  [(#895)](https://github.com/PennyLaneAI/catalyst/pull/895)

> * The `print_compilation_stage` function has been renamed `get_compilation_stage`.

Author: This was a bit intentional on my part. The other changelog entry initially worded `get_compilation_stage` as a modified `print_compilation_stage`, but I founded it a bit confusing so split it up into two entries (this one under breaking changes).

I think what I'll do is add in a mention of this breaking change in the entry above as well.

> * The `print_compilation_stage` function has been renamed `get_compilation_stage`.

Reviewer 2: Ah all good then 👌

> * The `print_compilation_stage` function has been renamed `get_compilation_stage`.

Author: Have added in a mention in the first changelog :+1:

<h3>Internal changes</h3>

* Adds `catalyst.from_plxpr.from_plxpr` for converting a PennyLane variant jaxpr into a
  Catalyst variant jaxpr.
  [(#837)](https://github.com/PennyLaneAI/catalyst/pull/837)

> * Catalyst is now compatible with Enzyme `v0.0.130`

Author: ```suggestion
* Catalyst now uses Enzyme `v0.0.130`
```

  [(#898)](https://github.com/PennyLaneAI/catalyst/pull/898)

* When memrefs have no identity layout, memrefs copy operations are replaced by the linalg copy operation.
  It does not use a runtime function but instead lowers to scf and standard dialects. It also ensures
  a better compatibility with Enzyme.
  [(#917)](https://github.com/PennyLaneAI/catalyst/pull/917)

> * llvm O2 and Enzyme passes are only run when needed (gradients presents). Async execution of QNodes

Reviewer 5: ```suggestion
* LLVM O2 and Enzyme passes are only run when needed (gradients presents). Async execution of QNodes
```

> * llvm O2 and Enzyme passes are only run when needed (gradients presents). Async execution of QNodes
>   triggers now triggers a Coroutine lowering pass.

Reviewer 2: Could be user facing in the sense that compilation time is significantly reduced when no gradients are used.
```suggestion
* LLVM's O2 optimization pipeline and Enzyme's AD transformations are now only run in the presence
  of gradients, significantly improving compilation times for programs without derivatives.
  Similarly, LLVM's coroutine lowering passes only run when `async_qnodes` is enabled in the QJIT decorator.
```

Author: Agree, I'll move this one into improvements.

* Catalyst's implementation of Lightning Kokkos plugin has been removed in favor of Lightning's one.
  [(#974)](https://github.com/PennyLaneAI/catalyst/pull/974)

> * Eliminate (some) scalar tensors from the IR by adding a `linalg-detensorize` pass at the end of the HLO lowering passes.

Reviewer 2: ```suggestion
* Eliminate some scalar tensors from the IR by adding a `linalg-detensorize` pass at the end of the HLO lowering pipeline.
  This will reduce memory usage at runtime as well reduce compilation complexity.
```
